### PR TITLE
misc: add a ScopedDict util, similar to LLVM's ScopedHashTable

### PR DESCRIPTION
**Pull request by superlopuh**

### CODE REVIEW

File: xdsl/utils/scoped_dict.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+from __future__ import annotations
+
+from typing import Generic, TypeVar
+
+_Key = TypeVar("_Key")
```
Comment (math-fehr) [inline xdsl/utils/scoped_dict.py:5]: Should this be hashable or something? I'm not sure why pyright doesn't complain here

File: xdsl/utils/scoped_dict.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    def __getitem__(self, key: _Key) -> _Value:
+        """
+        Fetch key from environment. Attempts to first fetch from current scope,
+        then from parent scopes. Raises KeyError error if not found.
+        """
+        local = self._local_scope.get(key)
+        if local is not None:
+            return local
+        if self.parent is None:
+            raise KeyError()
```
Comment (math-fehr) [inline xdsl/utils/scoped_dict.py:45]: Should this raise a better error message?

File: xdsl/utils/scoped_dict.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        return self.parent[key]
+
+    def __setitem__(self, key: _Key, value: _Value):
+        """
+        Assign key to current scope. Raises InterpretationError if key already
+        assigned to.
+        """
+        if key in self._local_scope:
+            raise ValueError(
```
Comment (math-fehr) [inline xdsl/utils/scoped_dict.py:54]: Should this be a `KeyError`?

**Review by math-fehr — APPROVED** [review]
👍

File: xdsl/utils/scoped_dict.py (hunk 3 above)
Comment (superlopuh) [reply xdsl/utils/scoped_dict.py:54]: I considered it, but I have a feeling that this is not the intended purpose of KeyError, which to me signals that something does not exist for a given key, whereas this error is raised when something does already exist. In either case, I'm not sure what the user can do to recover.

File: xdsl/utils/scoped_dict.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/utils/scoped_dict.py:5]: Hashable isn't a real type, AFAIK, and the `dict` generic type has the same `TypeVar` definition

**Pull request merged**